framework/src/kernels/MatBodyForce.C

```diff
@@ -0,0 +1,71 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MatBodyForce.h"
+#include "Function.h"
+
+registerMooseObject("MooseApp", MatBodyForce);
+registerMooseObject("MooseApp", ADMatBodyForce);
+
+template <bool is_ad, class Parent>
+InputParameters
+MatBodyForceTempl<is_ad, Parent>::validParams()
+{
+  InputParameters params = Parent::validParams();
+  params.addClassDescription("Kernel that defines a body force modified by a material property");
+  params.addRequiredParam<MaterialPropertyName>("material_property",
+                                                "Material property defining the property");
+  return params;
+}
+
+template <bool is_ad, class Parent>
+MatBodyForceTempl<is_ad, Parent>::MatBodyForceTempl(const InputParameters & parameters)
+  : Parent(parameters),
+    _property(this->template getGenericMaterialProperty<Real, is_ad>("material_property")),
+    _v_name(_var.name())
```

Comment:
Should this be in the nonad constructor?

Reply:
No, it doesn't have to be